soc: nxp: mcx: MCXN947: add NVS subsystem support with flash_k4 driver

DerekSnell opened the pull request:
- updates this SOC to use flash_k4 driver.  Uses flash_k4 driver for internal flash instead of ROM API driver.  One
benefit is the flash program phrase size decreases from 128 Bytes to 16 Bytes.  16 Byte phrases enables this SOC to leverage the Zephyr NVS subsystem, and the MCUboot swap mode.
- adds testcase for MCXN947 internal flash in samples/subsys/nvs.  For boards frdm_mcxn947 and mcx_n9xx_evks test NVS in internal flash in addition to external QSPI flash.  Also enable the NVS subsystem on these boards for Twister.

Commit:
boards: nxp: frdm_mcxn947 and mcx_n9xx_evk: enable clock for FMU

For internal flash FMU programg/erase operations, enable the FRO12MHZ
clock when the flash driver is enabled.

Signed-off-by: Derek Snell <derek.snell@nxp.com>

Commit:
samples: subsys: nvs: add testcase for MCXN947 internal flash

For boards frdm_mcxn947 and mcx_n9xx_evk test NVS in internal flash in
addition to external QSPI flash.  Also enable the NVS subsystem on
these boards for Twister.

Signed-off-by: Derek Snell <derek.snell@nxp.com>

DerekSnell commented:
@hakehuang , can you help test this?  This changes the driver used for internal flash.  Only the MCXN947 and MCXW devices use this flash_k4 driver currently.  The other MCXN and MCXA devices use the existing ROM API driver (for now).

hakehuang commented:
> @hakehuang , can you help test this? This changes the driver used for internal flash. Only the MCXN947 and MCXW devices use this flash_k4 driver currently. The other MCXN and MCXA devices use the existing ROM API driver (for now).

@DerekSnell , but this will cause the tests/drivers/flash cases failure on those board
```
2025-05-14 11:44:36,353 - twister - DEBUG - DEVICE: ===================================================================
2025-05-14 11:44:36,356 - twister - DEBUG - DEVICE: TESTSUITE flash_driver failed.
2025-05-14 11:44:36,360 - twister - DEBUG - DEVICE: ------ TESTSUITE SUMMARY START ------
2025-05-14 11:44:36,369 - twister - DEBUG - DEVICE: SUITE FAIL -  50.00% [flash_driver]: pass = 3, fail = 3, skip = 1, total = 7 duration = 2.213 seconds
2025-05-14 11:44:36,375 - twister - DEBUG - DEVICE: - PASS - [flash_driver.test_flash_copy] duration = 0.129 seconds
2025-05-14 11:44:36,381 - twister - DEBUG - DEVICE: - FAIL - [flash_driver.test_flash_erase] duration = 1.705 seconds
2025-05-14 11:44:36,386 - twister - DEBUG - DEVICE: - FAIL - [flash_driver.test_flash_fill] duration = 0.105 seconds
2025-05-14 11:44:36,392 - twister - DEBUG - DEVICE: - FAIL - [flash_driver.test_flash_flatten] duration = 0.104 seconds
2025-05-14 11:44:36,399 - twister - DEBUG - DEVICE: - PASS - [flash_driver.test_flash_page_layout] duration = 0.045 seconds
2025-05-14 11:44:36,404 - twister - DEBUG - DEVICE: - SKIP - [flash_driver.test_get_size] duration = 0.043 seconds
2025-05-14 11:44:36,411 - twister - DEBUG - DEVICE: - PASS - [flash_driver.test_read_unaligned_address] duration = 0.082 seconds
2025-05-14 11:44:36,415 - twister - DEBUG - DEVICE: ------ TESTSUITE SUMMARY END ------
2025-05-14 11:44:36,421 - twister - DEBUG - DEVICE: ===================================================================
2025-05-14 11:44:36,424 - twister - DEBUG - DEVICE: RunID: ff28e480f2713a77f8f07bd24c2121e6
2025-05-14 11:44:36,427 - twister - DEBUG - DEVICE: PROJECT EXECUTION FAILED
```

hakehuang commented on samples/subsys/nvs/sample.yaml:
```diff
@@ -21,3 +21,27 @@ tests:
            73 74 75 76 77 78 79 7a 7b 7c 7d 7e 7f"
+  sample.nvs.internal_flash:
+    tags: nvs
+    depends_on: nvs
```
we need add nvs in the <board>.yaml support list, otherwise this application will be filtered out.

DerekSnell replied:
Thanks @hakehuang ,
I already have `nvs` added to frdm_mcxn947_mcxn947_cpu0.yaml.  And I confirmed Twister will run this sample on these 2 boards.  Or are you referring to something else?

hakehuang replied:
Thanks. So only test on frdm board is enough.

butok commented on boards/nxp/frdm_mcxn947/frdm_mcxn947_mcxn947_cpu0.yaml:
```diff
@@ -24,6 +24,7 @@ supported:
   - i2c
   - i2s
   - i3c
+  - nvs
```
I am personally OK with using SW features, but Zephyr considers it as a mistake.
Please read the comment of @nashif:
https://github.com/zephyrproject-rtos/zephyr/pull/79924#issuecomment-2422424335

DerekSnell replied:
Thanks @butok .  But the NVS support does depend on the hardware.  NVS does not support flash with program phrase sizes over 32 Bytes.  This is the major reason this PR moves this SOC to the other driver, to enable NVS.  I feel we do need to filter the NVS tests and samples based on the HW platform.  NXP still has other platforms that are not compatible with NVS, like the LPC5500 family.

butok replied:
> But the NVS support does depend on the hardware.

Nice. I will use this argument in future discussions. ;)

DerekSnell commented:
> @DerekSnell , but this will cause the tests/drivers/flash cases failure on those board

@hakehuang , that test case fails from this known issue https://github.com/zephyrproject-rtos/zephyr/issues/88080.  And that test case tests the external QSPI flash, not the internal flash used with this PR.

hakehuang commented:
> @hakehuang , that test case fails from this known issue #88080. And that test case tests the external QSPI flash, not the internal flash used with this PR.

board testing for system/subsys/nvs pass on v4.1.0-3814-g72ab204c80ac

DerekSnell commented:
rebased to resolve merge conflicts.  Also added the MCXN547 SOC derivative from https://github.com/zephyrproject-rtos/zephyr/pull/89712

Commit:
soc: nxp: mcx: MCXNx4x: update SOC to use flash_k4 driver

Use flash_k4 driver for internal flash instead of ROM API driver.  One
benefit is the flash program phrase size decreases from 128 Bytes to 16
Bytes.  16 Byte phrases enables this SOC to leverage the Zephyr NVS
subsystem, and the MCUboot swap mode.

Signed-off-by: Derek Snell <derek.snell@nxp.com>

Conflicts:
	west.yml

hakehuang commented:
board testing pass for 
•	board_list :['frdm_mcxn947/mcxn947/cpu0', 'mcx_n9xx_evk/mcxn947/cpu0']
•	commit_id :96a637c2aad472a8eaca077674b9a4b35e59335e
•	pull_id :89909
•	version :v4.1.0-4937-g96a637c2aad4

dleach02 commented:
@DerekSnell your HAL PR was merged so I updated this one to have the new SHA
